WordPress/gutenberg
# Cover block: background element's classname consistency

## Description

Following on from conversations in: https://github.com/WordPress/gutenberg/pull/38362#issuecomment-1026311465

Here is a Pull Request that updates the classnames on the Cover Block background span element in order to:

1. Remove duplicated `gradientClass` where there is no media url
2. Add a permanent classname to the background `<span  />` of `wp-block-cover__background` to replace `wp-block-cover__gradient-background`, since the former latter applied to the element all the time (even where there is no gradient)
3. Revert the addition of `wp-block-cover__gradient-background`  to the logic before the the [change in October 2021](https://github.com/WordPress/gutenberg/commit/b5a632980cf2efed6ee998da0422b965fc447f38#diff-c9b6422f332979151b7228eb2af812e84859338e6b4d5ef38e5341ef6e8f5a39R695)
4. ~Achieve World Peace ☮️~ 

So, in the save.js, the element is switching from:

```jsx
				<span
					aria-hidden="true"
					className={ classnames(
						overlayColorClass,
						dimRatioToClass( dimRatio ),
						'wp-block-cover__gradient-background',
						gradientClass,
						{
							'has-background-dim': dimRatio !== undefined,
							'has-background-gradient':
								gradient || customGradient,
							[ gradientClass ]: ! url && gradientClass,
						}
					) }
					style={ bgStyle }
				/>
```

To:

```jsx
			<span
				aria-hidden="true"
				className={ classnames(
					'wp-block-cover__background',
					overlayColorClass,
					dimRatioToClass( dimRatio ),
					{
						'has-background-dim': dimRatio !== undefined,
						// For backwards compatibility. Former versions of the Cover Block applied
						// `.wp-block-cover__gradient-background` in the presence of
						// media, a gradient and a dim.
						'wp-block-cover__gradient-background':
							url && gradientValue && dimRatio !== 0,
						'has-background-gradient': gradientValue,
						[ gradientClass ]: gradientClass,
					}
				) }
				style={ bgStyle }
			/>
```

We've added an extra deprecation and updated the fixtures and other tests.

## Testing Instructions

Here are some test blocks created using trunk. Paste them in the code view of the block editor and ensure that there are no validation errors.

<details><summary>Test blocks</summary>

```html

<div class="wp-block-cover is-light"><span aria-hidden="true" class="wp-block-cover__gradient-background has-background-dim"></span><img class="wp-block-cover__image-background wp-image-216" alt="" src="http://localhost:4759/wp-content/uploads/2022/01/puppy.jpg" data-object-fit="cover"/><div class="wp-block-cover__inner-container">
<p class="has-text-align-center has-large-font-size">Cover block with image and dim</p>
</div></div>



<div class="wp-block-cover"><span aria-hidden="true" class="has-background-dim-0 wp-block-cover__gradient-background has-pale-ocean-gradient-background has-background-dim has-background-gradient"></span><img class="wp-block-cover__image-background wp-image-213" alt="" src="http://localhost:4759/wp-content/uploads/2022/01/how-to-grow-organic-fiddlehead-ferns-2539638-hero-23c689cdd2b74f0c9e817cc6e710f0b8-scaled.jpg" data-object-fit="cover"/><div class="wp-block-cover__inner-container">
<p class="has-text-align-center has-large-font-size">Cover block with image, gradient and no dim</p>
</div></div>



<div class="wp-block-cover"><span aria-hidden="true" class="has-background-dim-80 wp-block-cover__gradient-background has-cool-to-warm-spectrum-gradient-background has-background-dim has-background-gradient"></span><img class="wp-block-cover__image-background wp-image-213" alt="" src="http://localhost:4759/wp-content/uploads/2022/01/how-to-grow-organic-fiddlehead-ferns-2539638-hero-23c689cdd2b74f0c9e817cc6e710f0b8-scaled.jpg" data-object-fit="cover"/><div class="wp-block-cover__inner-container">
<p class="has-text-align-center has-large-font-size">Cover block with image, dim, and custom gradient</p>
</div></div>



<div class="wp-block-cover is-light"><span aria-hidden="true" class="has-highlight-background-color has-background-dim-100 wp-block-cover__gradient-background has-background-dim"></span><div class="wp-block-cover__inner-container">
<p class="has-text-align-center has-large-font-size">Cover block with plain background</p>
</div></div>



<div class="wp-block-cover is-light"><span aria-hidden="true" class="has-background-dim-100 wp-block-cover__gradient-background has-luminous-dusk-gradient-background has-background-dim has-background-gradient has-luminous-dusk-gradient-background"></span><div class="wp-block-cover__inner-container">
<p class="has-text-align-center has-large-font-size">Cover block with gradient background</p>
</div></div>


```
</details>


## Types of changes
Block classname changes and deprecation

## Checklist:
